ARCs/arc-0060.md

```diff
@@ -0,0 +1,328 @@
+
+ARC-1 defines a standard for signing Algorand transactions that are represented as structured objects. This proposal extends the signing process to byte arrays encoded with a standardized structure.
+
+[ARC-60](./arc-0060.md) defines an API for wallets to sign structured data that are not Algorand transactions.
+
+## Motivation
+
+Signing data is a common and critical operation. Users may need to sign data for multiple reasons (e.g. delegate signatures, DIDs, signing documents, authentication).
+
+Algorand wallets need a standard approach to byte signing to unlock self-custodial services and protect users from malicious and attack-prone signing workflows.
+
+This ARC provides a standard API for bytes signing. The API encodes byte arrays into well structured JSON schemas together with additional metadata. It requires wallets to validate the signing inputs, notify users about what they are signing and warn them in case of dangerous signing requests.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative
+
+### Overview
+
+> This section is non-normative
+
+Signatures of bytes are processed with the `signData(data, metadata, signer)` function.
+
+`data` is a string representing the structured data being signed. It is the representation of a JSON Schema, provided with the `metadata`.
+
+`metadata` is a `StdSignMetadata` object that describes the signature scope and the type of data being signed, including the encoding and the JSON schema.
+
+`signer` is a byte array representing the public key to use for the signing operation.
+
+### Interfaces
+
+> Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects.
+
+ARC-60 uses interchangeably the terms "throw an error" and "reject a promise with an error".
+
+#### **Interface `SignDataFunction`**
+
+A wallet function `signData` is defined by the interface:
+
+```tsx
+export type SignDataFunction = {
+    data: StdDataStr,
+    metadata: StdSignMetadata,
+    signer: Ed25519Pk,
+}
+    => Promise<(SignedData | null)>;
+```
+
+- `data` is a `StdDataStr` object representing the structured data being signed (defined below).
+- `metadata` is a `StdSignMetadata` objects (defined below) that provide additional information on the data being signed, such as the `data` encoding and the JSON schema.
+- `signer` is a `Ed25519Pk` object (defined below) that represents the signer public key.
+
+The `signData` function returns a `SignedData` object or, in case of error, rejects the promise with an error object `SignDataError`.
+
+#### Interface `StdDataStr`
+
+The `StdDataStr` object is a string that comply with the `metadata` JSON schema.
+
+```tsx
+export type StdDataStr = string;
```

Comment:
I don't understand how this can contain an logicsig program if it's a string

Reply:
Yes, I prefer a Buffer or Uint8Array.

Reply:
The way I read the spec is that the StdDataStr is a stringified JSON object that must comply with the metadata supplied as an argument to sign function, of type StdSignMetadata. The metadata specifies which field contains data to be signed, and not the whole json object. Part of the problem is this line in the first paragraphs:  

> data is a string representing the structured data being signed. 

This is not true according to the rest of the spec. The data being signed is whichever field of the StdDataStr json to be signed. 

Another problem is that there is not defined here the standard encoding for JSON bytes data. Typically in web applications it is Base64 encoded strings, and OpenAPI spec clients/servers will work with that, but it's only by common convention, and I don't know the base specs like JSON Schema say anything about it (off the top of my head)

In general I find the spec confusing because of a) the use of both TypeScript types and JSON types, and I don't like the use of things like UInt8Array at all, as they are TS domain specific concepts and don't belong in language agnostic specs. b) slightly misleading terminology, like "Scope" instead of "Schema"...

Reply:
Thank you @FrankSzendzielarz 

This commit https://github.com/algorandfoundation/ARCs/pull/284/commits/c08b9958f339073ce1711f080e7e34a7af3b940e should fix the issue you raised.

Reply:
Also, @jasonpaulos @mxmauro please refer to the [Backwards Compatibility](https://github.com/deanstef/ARCs/blob/byte-signing/ARCs/arc-0060.md#backwards-compatibility) section of the ARC.